### PR TITLE
[Feature] Support to set data root through commands

### LINKED ISSUE #6915: Is there a better way to specify data root?
Hi! I was wondering if there's a better way to change the data root. Specifically, when I tried to train on COCO I specified `data_root` either through `--cfg-options` or hard-code it in the config file. However, only `data_root` was successfully changed whereas `img_prefix`, `ann_file` under `cfg.data.train` and `cfg.data.val` are still populated by the default `data/coco/` (for example, `data/coco/train2017/`). It looks like those fields have already been populated in `_base_/datasets/coco_detection.py` before my change to `data_root` takes place.

I want to avoid making symlinks under the current project. How to overwrite fields like this?

Thank you!

### PR DESCRIPTION
**Pull request by CCODING04**
## Motivation

Support to set data root throught commands like export `MMDET_DATASETS=/path/to/datasets`

## Modification

update `tools/train.py`  
```python
# line 111
# update data root according to environment variable
    if os.environ.get('MMDET_DATASETS', None) is not None:
        cfg.data_root = os.environ['MMDET_DATASETS']
```
- demo
```bash
MMDET_DATASETS=/path/to/datasets python tools/train.py \
${CONFIG_FILE} \
[optional arguments]
```

### CODE REVIEW

**Comment by hhaAndroid** [comment]
@CCODING04 This way of writing may be wrong, you can run a simple experiment to see.

**Comment by CCODING04** [comment]
@hhaAndroid  
Seems like I push some commits irregularly, but can you review my new code ? 
I have tested it and it works.

**Comment by jbwang1997** [comment]
It seems due to I rebase the branch. now it is normal.

**Comment by CCODING04** [comment]
> It seems due to I rebase the branch. now it is normal.

File: tools/train.py [hunk 1]
```diff
@@ -98,6 +98,17 @@ def main():
...
 
     cfg = Config.fromfile(args.config)
+
+    # update data root according to environment variable
+    if os.environ.get('MMDET_DATASETS', None) is not None:
+        def update_data_root(cfg, str_o, str_n):
+            for k, v in cfg.items():
+                if isinstance(v, mmcv.ConfigDict):
+                    update_data_root(cfg[k], str_o, str_n)
+                if isinstance(v, str):
```
Comment (hhaAndroid) [inline tools/train.py:108]: if isinstance(v, str) and str_o in v:

File: tools/train.py [hunk 2]
```diff
@@ -98,6 +98,17 @@ def main():
     args = parse_args()
 
     cfg = Config.fromfile(args.config)
+
+    # update data root according to environment variable
+    if os.environ.get('MMDET_DATASETS', None) is not None:
```
Comment (hhaAndroid) [inline tools/train.py:103]: This logic needs to be written as a separate function, because `test.py` and`browse_dataset.py` also need.

Comment (hhaAndroid) [reply tools/train.py:103]: After that, add unit tests to ensure correctness. Thank you !

File: tools/train.py (hunk 1 above)
Comment (CCODING04) [reply tools/train.py:108]: 😞  fix it

File: tools/train.py (hunk 2 above)
Comment (CCODING04) [reply tools/train.py:103]: > After that, add unit tests to ensure correctness. Thank you !

@hhaAndroid 😸 Okay~~

File: tools/train.py [hunk 3]
```diff
@@ -98,6 +98,17 @@ def main():
...
+        def update_data_root(cfg, str_o, str_n):
+            for k, v in cfg.items():
+                if isinstance(v, mmcv.ConfigDict):
+                    update_data_root(cfg[k], str_o, str_n)
+                if isinstance(v, str):
+                    cfg[k] = v.replace(str_o, str_n)
+        update_data_root(cfg, cfg.data_root, os.environ['MMDET_DATASETS'])
```
Comment (jbwang1997) [inline tools/train.py:110]: We can only replace the path in cfg.data.

File: tools/train.py (hunk 2 above)
Comment (jbwang1997) [inline tools/train.py:103]: If `MMDET_DATASETS` is not None, we need to print some information in the log.

**Comment by jbwang1997** [comment]
And we can also add some descriptions of this function in [docs/en/1_exist_data_model](https://github.com/open-mmlab/mmdetection/blob/master/docs/en/1_exist_data_model.md) and [docs/zh_cn/1_exist_data_model](https://github.com/open-mmlab/mmdetection/blob/master/docs/zh_cn/1_exist_data_model.md)

**Comment by CCODING04** [comment]
@hhaAndroid @jbwang1997 
Hi, I have finished the code. Plz check it. Thanks 😺 
If everything is ok, I  will add detailed desc in doc later.

File: tools/utils.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
+# Copyright (c) OpenMMLab. All rights reserved.
```
Comment (hhaAndroid) [inline tools/utils.py:1]: Please place to https://github.com/open-mmlab/mmdetection/blob/master/mmdet/utils/misc.py

File: tools/train.py [hunk 5]
```diff
@@ -98,6 +99,12 @@ def main():
     args = parse_args()
 
     cfg = Config.fromfile(args.config)
+
+    # update data root according to MMDET_DATASETS
+    if os.environ.get('MMDET_DATASETS', None) is not None:
```
Comment (hhaAndroid) [inline tools/train.py:104]: I recommend moving into the function as well

File: tools/utils.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+import mmcv
+from mmcv.utils import print_log
+
+def update_data_root(cfg, data_root_n, logger=None):
+    """updata data root
```
Comment (hhaAndroid) [inline tools/utils.py:6]: The content of docs is relatively small, can you describe more?

Comment (hhaAndroid) [reply tools/utils.py:6]: data_root_n -> dst_root

File: tools/utils.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    Args:
+        cfg (mmcv.Config): model config
+        data_root_n (str): new data root
+        logger (logging.Logger | str | None): the way to print msg
+
+    """
+    assert isinstance(cfg, mmcv.Config), \
+        f"cfg got wrong type: {type(cfg)}, expected mmcv.Config"
+    
+    def update(cfg, str_o, str_n):
```
Comment (hhaAndroid) [inline tools/utils.py:17]:
```suggestion
    def update(cfg, src_str, dst_str):
```

**Comment by CCODING04** [comment]
@hhaAndroid 
Got it, I will do it later, thanks for your guidance! Learn a lot! 😄

**Comment by CCODING04** [comment]
@hhaAndroid 
Hi, I refactor the code as your suggestion. Plz review it. ♨️

File: mmdet/utils/misc.py [hunk 8]
```diff
@@ -36,3 +39,40 @@ def find_latest_checkpoint(path, suffix='pth'):
...
+    
+    def update(cfg, src_str, dst_str):
+        for k, v in cfg.items():
+            if isinstance(v, mmcv.ConfigDict):
+                update(cfg[k], src_str, dst_str)
+            if isinstance(v, str) and src_str in v:
+                cfg[k] = v.replace(src_str, dst_str)
+
+    update(cfg.data, cfg.data_root, dst_root)
+    cfg.data_root = dst_root
```
Comment (ZwwWayne) [inline mmdet/utils/misc.py:78]: The code format is incorrect. Please use pre-commit

Comment (CCODING04) [reply mmdet/utils/misc.py:78]: > The code format is incorrect. Please use pre-commit

Sry, I fogot it. Try it now.

Comment (CCODING04) [reply mmdet/utils/misc.py]: @ZwwWayne 
Done pre-commit alredy. Plz check.

File: tools/misc/browse_dataset.py [hunk 9]
```diff
@@ -55,6 +56,10 @@ def skip_pipeline_steps(config):
         ]
 
     cfg = Config.fromfile(config_path)
+
+    # update data root according to MMDET_DATASETS
+    update_data_root(cfg)
```
Comment (hhaAndroid) [inline tools/misc/browse_dataset.py:61]: Also need to modify analyze_results, benchmark, confusion_matrix, eval_metric, optimize_anchors and print_config

File: mmdet/utils/misc.py [hunk 10]
```diff
@@ -1,8 +1,12 @@
 # Copyright (c) OpenMMLab. All rights reserved.
 import glob
+import os
 import os.path as osp
 import warnings
 
+import mmcv
```
Comment (hhaAndroid) [inline mmdet/utils/misc.py:7]: Please write docs about this new feature [here](https://github.com/open-mmlab/mmdetection/blob/master/docs/en/3_exist_data_new_model.md#prepare-the-standard-dataset)

File: mmdet/utils/misc.py [hunk 11]
```diff
@@ -36,3 +40,38 @@ def find_latest_checkpoint(path, suffix='pth'):
...
+    If set env MMDET_DATASETS, update cfg.data_root according to
+    MMDET_DATASETS. Otherwise, using cfg.data_root as default.
+
+    Args:
+        cfg (mmcv.Config): The model config need to modify
+        logger (logging.Logger | str | None): the way to print msg
+    """
+    if 'MMDET_DATASETS' not in os.environ:
+        print_log(
+            f'Not set env MMDET_DATASETS, using {cfg.data_root} as data root.',
```
Comment (ZwwWayne) [inline mmdet/utils/misc.py:57]: Not set env MMDET_DATASETS -> MMDET_DATASETS is not set in environments,

File: mmdet/utils/misc.py [hunk 12]
```diff
@@ -36,3 +40,38 @@ def find_latest_checkpoint(path, suffix='pth'):
...
+            logger)
+        return None
+    else:
+        dst_root = os.environ['MMDET_DATASETS']
+        print_log(
+            f'Set env MMDET_DATASETS={dst_root}, \
```
Comment (ZwwWayne) [inline mmdet/utils/misc.py:63]: Using MMDET_DATASETS=${dst_root} as datat root.

File: mmdet/utils/misc.py [hunk 13]
```diff
@@ -36,3 +40,38 @@ def find_latest_checkpoint(path, suffix='pth'):
...
+
+    If set env MMDET_DATASETS, update cfg.data_root according to
+    MMDET_DATASETS. Otherwise, using cfg.data_root as default.
+
+    Args:
+        cfg (mmcv.Config): The model config need to modify
+        logger (logging.Logger | str | None): the way to print msg
+    """
+    if 'MMDET_DATASETS' not in os.environ:
+        print_log(
```
Comment (hhaAndroid) [inline mmdet/utils/misc.py:56]: I don't think it needs to be printed because nothing has changed. Avoid misleading users

**Comment by CCODING04** [comment]
@ZwwWayne @hhaAndroid 
Update `misc.py`, please review, thanks!

File: mmdet/utils/misc.py [hunk 14]
```diff
@@ -36,3 +40,34 @@ def find_latest_checkpoint(path, suffix='pth'):
...
+    """Update data root according to env MMDET_DATASETS.
+
+    If set env MMDET_DATASETS, update cfg.data_root according to
+    MMDET_DATASETS. Otherwise, using cfg.data_root as default.
+
+    Args:
+        cfg (mmcv.Config): The model config need to modify
+        logger (logging.Logger | str | None): the way to print msg
+    """
+    if 'MMDET_DATASETS' not in os.environ:
```
Comment (hhaAndroid) [inline mmdet/utils/misc.py:55]:
```suggestion
   assert isinstance(cfg, mmcv.Config), \
        f'cfg got wrong type: {type(cfg)}, expected mmcv.Config'
        
    if 'MMDET_DATASETS'  in os.environ:
         dst_root = os.environ['MMDET_DATASETS']
         print_log(f'MMDET_DATASETS has been set to be {dst_root}. \
                Using {dst_root} as data root.')
```

**Comment by hhaAndroid** [comment]
@CCODING04 Please fix the conflict as we plan to merge into the dev branch.

File: mmdet/utils/misc.py [hunk 15]
```diff
@@ -36,3 +40,37 @@ def find_latest_checkpoint(path, suffix='pth'):
...
+    """Update data root according to env MMDET_DATASETS.
+
+    If set env MMDET_DATASETS, update cfg.data_root according to
+    MMDET_DATASETS. Otherwise, using cfg.data_root as default.
+
+    Args:
+        cfg (mmcv.Config): The model config need to modify
+        logger (logging.Logger | str | None): the way to print msg
+    """
+    assert isinstance(cfg, mmcv.Config), \
```
Comment (hhaAndroid) [inline mmdet/utils/misc.py:55]: It's my mistake, please fix it as follows

```python
assert isinstance(cfg, mmcv.Config), \ ...

if 'MMDET_DATASETS'  not in os.environ:
    return

dst_root = os.environ['MMDET_DATASETS']
print_log(f'MMDET_DATASETS has been set to be {dst_root}.'
        f'Using {dst_root} as data root.')

def update(cfg, src_str, dst_str):
...
```

Comment (CCODING04) [reply mmdet/utils/misc.py:55]: @hhaAndroid
Thank you a lot !
Sorry not to review the code carefully and make this stupid mistake, I'll check it more carefully before commit next time.

File: docs/en/3_exist_data_new_model.md [hunk 16]
```diff
@@ -41,6 +41,12 @@ mmdetection
 
 ```
 
+Or you can set your dataset root through
+```bash
```
Comment (ZwwWayne) [inline docs/en/3_exist_data_new_model.md:45]: Need an empty line before and after the code block, according to the markdown style.

Comment (CCODING04) [reply docs/en/3_exist_data_new_model.md:45]: > Need an empty line before and after the code block, according to the markdown style.

Thanks for your correction, I'll take care of it next time.
Thanks MMLAB, it's my first time to participate in open source project, and also my first PR.
Thanks @hhaAndroid @ZwwWayne  again, I'll review the mistake I took and be careful next time.
Thanks :)

**Pull request merged**